File: deploy/deploy.ts
import { ethers } from "ethers";
import { Verifier__factory, Kaviar__factory } from "../types";
import {
  parse_chain_params,
  parse_chain_provider,
  chains,
} from "../utils/parseChainUtils";

async function main() {
  const readline = require("node:readline").createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  readline.question("Enter Chain : ", (chain_name: string) => {
    let chain = chain_name;

    try {
      if (chain == "") {
        throw new Error("Chain must be set");
      } else {
        if (!chains.includes(chain)) {
          throw new Error("Chain could not be parsed because of invalid name");
        }
      }
    } catch (e) {
      console.log(e);
      readline.close();
      return;
    }

    run(chain);
    readline.close();
  });
}

async function run(chain: string) {
  const chain_params = parse_chain_params(chain);

  // deploy KAVIAR on chain
  const wallet = new ethers.Wallet(process.env.userOldSigner ?? "");
  const provider = parse_chain_provider(chain);
  const signer = wallet.connect(provider);

  console.log(`\nDeploying contract Verifier.sol on ${chain}\n`);

  const verifier = await new Verifier__factory(signer).deploy();
  await verifier.deployed();
  console.log(
    `Verifier.sol has been deployed to ${verifier.address} on ${chain}\n`
  );

  const ETH_AMOUNT = ethers.utils.parseEther("0.0001");

  // deploy KAVIAR on chain
  console.log(`Deploying contract Kaviar.sol on ${chain}\n`);

  const kaviar = await new Kaviar__factory(signer).deploy(
    verifier.address,
    ETH_AMOUNT,
    chain_params.poseidon
  );
  await kaviar.deployed();
  console.log(`Kaviar.sol has been deployed to ${kaviar.address} on ${chain}`);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: src/deploy.rs
use crate::{Config, Kaviar, Verifier};

use dotenv::dotenv;
use ethers::{
    middleware::SignerMiddleware,
    prelude::*,
    providers::{Http, Provider},
    signers::{LocalWallet, Signer},
    utils,
};
use eyre::Result;
use std::{convert::TryFrom, sync::Arc};

pub async fn deploy(config: Config) -> Result<(), &'static str> {
    dotenv().ok();

    match config {
        Config::Deploy { chain, chain_name } => {
            let user: &str = &std::env::var("userOldSigner").expect("User Old Signer must be set");
            // Deploy Verifier.sol
            let url: &str = &std::env::var(&chain.chain_env_var).expect("URL must be set");
            let provider = Provider::<Http>::try_from(url).expect("Failed to create provider");
            let wallet: LocalWallet = user
                .parse::<LocalWallet>()
                .expect("Failed to parse wallet")
                .with_chain_id(chain.chain_enum_value);
            let client_verifier = Arc::new(SignerMiddleware::new(provider.clone(), wallet.clone()));

            println!("\nDeploying contract Verifier.sol on {}\n", chain_name);
            let verifier_contract = Verifier::deploy(client_verifier, ())
                .unwrap()
                .send()
                .await
                .unwrap();

            let addr_verifier = verifier_contract.address();
            println!(
                "Verifier.sol has been deployed to {:?} on {}\n",
                addr_verifier, chain_name
            );

            // Deploy Kaviar.sol
            let client_kaviar = Arc::new(SignerMiddleware::new(provider.clone(), wallet.clone()));

            let poseidon = chain.poseidon.parse::<Address>().expect("Invalid Address");

            // constants
            let eth_amount = utils::parse_ether(0.01).expect("Failed to parse Ether");

            println!("Deploying contract Kaviar.sol on {}\n", chain_name);
            let kaviar_contract =
                Kaviar::deploy(client_kaviar, (addr_verifier, eth_amount, poseidon))
                    .unwrap()
                    .send()
                    .await
                    .unwrap();

            let addr_kaviar = kaviar_contract.address();
            println!(
                "Kaviar.sol has been deployed to {:?} on {}\n",
                addr_kaviar, chain_name
            );
            Ok(())
        }
        _ => Err("Invalid Params"),
    }
}
